### PR TITLE
Fix redirecting StandardError stream

### LINKED ISSUE #33761: Missing StandardError stream redirect
```
System.InvalidOperationException: StandardError has not been redirected.
   at System.Diagnostics.Process.BeginErrorReadLine()
   at System.CodeDom.Compiler.Executor.ExecWaitWithCapture(IntPtr userToken, String cmd, String currentDir, TempFileCollection tempFiles, String& outputName, String& errorName)
```

### PR DESCRIPTION
**Pull request by bazuka5801**
Fixes #33761

### CODE REVIEW

**Comment by bazuka5801** [comment]
Thanks for approving!)
But I don't have tests.

**Comment by stephentoub** [comment]
> But I don't have tests.

Could you write one and include it in the PR?

**Comment by bazuka5801** [comment]
> > But I don't have tests.
> 
> Could you write one and include it in the PR?

I doubt that I will succeed

File: src/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs [hunk 1]
```diff
@@ -62,7 +62,8 @@ public static int ExecWaitWithCapture(IntPtr userToken, string cmd, string curre
                 {
                     WorkingDirectory = currentDir,
                     RedirectStandardOutput = true,
-                    RedirectStandardInput = true
+                    RedirectStandardInput = true,
```
Comment (stephentoub) [inline src/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs]: Actually, this Input should be replaced with Error.

**Comment by karelz** [comment]
@krwq can you provide guidance / pointers how to test it?
Unless there are special circumstances (e.g. impossible to test scenario), we should have test as part of the PR ...

File: src/System.CodeDom/tests/System/CodeDom/Compiler/ExecutorTests.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        public void ExecWaitWithCapture_FilesSpecified_OutputCaptured(bool specifyOutput, bool specifyError)
+        {
+            var tfc = new TempFileCollection(TestDirectory);
+
+            string outputName = specifyOutput ? GetTestFilePath() : null;
+            string errorName = specifyError ? GetTestFilePath() : null;
+
+            Executor.ExecWaitWithCapture(s_cmd, tfc, ref outputName, ref errorName);
+
+            Assert.Equal(
```
Comment (ahsonkhan) [inline src/System.CodeDom/tests/System/CodeDom/Compiler/ExecutorTests.cs]: nit: This assert is a bit difficult to reason about. You already have asserts within `!specifyOutput` and `!specifyError` blocks. Add the four branches and asserts there.

Comment (stephentoub) [reply src/System.CodeDom/tests/System/CodeDom/Compiler/ExecutorTests.cs]: I simplified.

**Comment by stephentoub** [comment]
@dotnet-bot test OSX x64 Debug Build please
@dotnet-bot test Linux arm64 Release Build please

**Comment by stephentoub** [comment]
@dotnet-bot test this plesae

**Comment by stephentoub** [comment]
@dotnet-bot test this plesae

**Comment by stephentoub** [comment]
@dotnet/dnceng, is there an issue with Jenkins? Doesn't appear to be responding to test requests.

File: src/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs [hunk 3]
```diff
@@ -91,6 +91,8 @@ public static int ExecWaitWithCapture(IntPtr userToken, string cmd, string curre
                         throw new ExternalException(SR.Format(SR.ExecTimeout, cmd), WAIT_TIMEOUT);
                     }
 
+                    p.WaitForExit();
```
Comment (krwq) [inline src/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs]: why is this needed? we already wait for process to exit line above

Comment (stephentoub) [reply src/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs]: See the docs for WaitForExit(int).

**Comment by stephentoub** [comment]
@mmitche, I see you closed and reopened the PR. Any idea why legs aren't restarting, e.g. the Linux x64 release leg?

File: src/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs (hunk 3 above)
Comment (krwq) [reply src/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs]: per docs, it returns true when process did exit and then we wait again, when it is false we throw `ExternalException`. Am I missing something?

Comment (krwq) [reply src/System.CodeDom/src/System/CodeDom/Compiler/Executor.cs]: Ah I see.. "To ensure that asynchronous event handling has been completed, call the WaitForExit() overload that takes no parameter after receiving a true from this overload"

**Comment by mmitche** [comment]
@stephentoub They all did.  I think you needed a refresh

**Comment by stephentoub** [comment]
> They all did. I think you needed a refresh

I'm not seeing that.  But regardless, do you know why all of the tests in this PR and every other one in corefx are hanging?  It's not the tests themselves: all suites across multiple legs are hanging before any tests have actually been run, and then eventually timeout after 12 hours.  For example, here's one that's been going for 10 hours:
https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/19120/

**Comment by mmitche** [comment]
@stephentoub No idea.  @MattGal will need to take a look in the morning.

**Comment by MattGal** [comment]
The issue is https://github.com/dotnet/core-eng/issues/4729 which we've attempted to mitigate, and which _might_ be resolved around 3 PM.

**Comment by stephentoub** [comment]
@dotnet-bot test this please
